File: packages/core/src/ml/reduce.ts
/**
 * Pure, dependency-free dimensionality reduction for embedding projections.
 * {@link pca} projects an `n×d` matrix onto its top-`k` principal components via
 * covariance power-iteration with deflation — deterministic (no RNG), so tests
 * and re-renders are stable. Pair with {@link addEmbedding}. For t-SNE/UMAP,
 * feed precomputed 2-D coordinates straight into the builder.
 */

/** The result of {@link pca}: projected scores + the components they project onto. */
export interface PcaResult {
  /** Projected coordinates, row-major `n×k`. */
  scores: Float64Array;
  /** Principal-component directions, row-major `k×d` (unit vectors). */
  components: Float64Array;
  /** Explained-variance ratio per component, length `k`. */
  explained: Float64Array;
  /** Per-column mean subtracted before projection, length `d`. */
  mean: Float64Array;
  n: number;
  d: number;
  k: number;
}

/** Z-score each of the `d` columns of a row-major `n×d` matrix (unit variance). */
export function standardize(data: ArrayLike<number>, n: number, d: number): Float64Array {
  const out = new Float64Array(n * d);
  for (let j = 0; j < d; j++) {
    let mean = 0;
    for (let i = 0; i < n; i++) mean += data[i * d + j]!;
    mean /= Math.max(1, n);
    let varsum = 0;
    for (let i = 0; i < n; i++) { const dv = data[i * d + j]! - mean; varsum += dv * dv; }
    const sd = Math.sqrt(varsum / Math.max(1, n - 1)) || 1;
    for (let i = 0; i < n; i++) out[i * d + j] = (data[i * d + j]! - mean) / sd;
  }
  return out;
}

/**
 * Principal component analysis of a row-major `n×d` matrix, projected to `k`
 * dims (default 2). Centers columns, forms the `d×d` covariance, then extracts
 * the top-`k` eigenvectors by power iteration + deflation. Deterministic.
 */
export function pca(data: ArrayLike<number>, n: number, d: number, k = 2): PcaResult {
  k = Math.max(1, Math.min(k, d));
  const mean = new Float64Array(d);
  for (let i = 0; i < n; i++) for (let j = 0; j < d; j++) mean[j]! += data[i * d + j]!;
  for (let j = 0; j < d; j++) mean[j]! /= Math.max(1, n);

  // Centered copy, then covariance = Xcᵀ·Xc / (n−1) (symmetric d×d).
  const Xc = new Float64Array(n * d);
  for (let i = 0; i < n; i++) for (let j = 0; j < d; j++) Xc[i * d + j] = data[i * d + j]! - mean[j]!;
  const cov = new Float64Array(d * d);
  const denom = Math.max(1, n - 1);
  for (let i = 0; i < n; i++) {
    const row = i * d;
    for (let a = 0; a < d; a++) {
      const xa = Xc[row + a]!;
      if (xa === 0) continue;
      for (let b = a; b < d; b++) cov[a * d + b]! += xa * Xc[row + b]!;
    }
  }
  let totalVar = 0;
  for (let a = 0; a < d; a++) {
    for (let b = a; b < d; b++) { const v = cov[a * d + b]! / denom; cov[a * d + b] = v; cov[b * d + a] = v; }
    totalVar += cov[a * d + a]!;
  }

  const components = new Float64Array(k * d);
  const explained = new Float64Array(k);
  const v = new Float64Array(d), w = new Float64Array(d);
  for (let c = 0; c < k; c++) {
    // Deterministic non-degenerate seed, orthogonal-ish across components.
    for (let j = 0; j < d; j++) v[j] = Math.sin(1 + j * (c + 1)) + 0.5;
    normalize(v, d);
    let eig = 0;
    for (let iter = 0; iter < 256; iter++) {
      for (let a = 0; a < d; a++) { let s = 0; const ra = a * d; for (let b = 0; b < d; b++) s += cov[ra + b]! * v[b]!; w[a] = s; }
      const norm = normalize(w, d);
      let dot = 0;
      for (let a = 0; a < d; a++) dot += w[a]! * v[a]!;
      for (let a = 0; a < d; a++) v[a] = w[a]!;
      if (Math.abs(Math.abs(dot) - 1) < 1e-9) { eig = norm; break; }
      eig = norm;
    }
    for (let j = 0; j < d; j++) components[c * d + j] = v[j]!;
    explained[c] = totalVar > 0 ? eig / totalVar : 0;
    // Deflate: cov ← cov − eig·v·vᵀ so the next pass finds the next component.
    for (let a = 0; a < d; a++) { const va = eig * v[a]!; for (let b = 0; b < d; b++) cov[a * d + b]! -= va * v[b]!; }
  }

  // scores = Xc · componentsᵀ  (n×k)
  const scores = new Float64Array(n * k);
  for (let i = 0; i < n; i++) {
    const row = i * d;
    for (let c = 0; c < k; c++) {
      let s = 0; const cr = c * d;
      for (let j = 0; j < d; j++) s += Xc[row + j]! * components[cr + j]!;
      scores[i * k + c] = s;
    }
  }
  return { scores, components, explained, mean, n, d, k };
}

/** L2-normalize `vec` in place; returns the original norm (0 → left as zeros). */
function normalize(vec: Float64Array, d: number): number {
  let s = 0;
  for (let j = 0; j < d; j++) s += vec[j]! * vec[j]!;
  const norm = Math.sqrt(s);
  if (norm > 0) for (let j = 0; j < d; j++) vec[j]! /= norm;
  return norm;
}

File: native/src/ml/reduce.cpp
#include "ml/reduce.hpp"

#include <algorithm>
#include <cmath>

namespace photon::ml {

namespace {

/// L2-normalise in place; returns the norm it had. A zero vector is left alone.
double normalize(std::vector<double>& vec, size_t d) {
  double s = 0.0;
  for (size_t j = 0; j < d; ++j) s += vec[j] * vec[j];
  const double norm = std::sqrt(s);
  if (norm > 0.0) {
    for (size_t j = 0; j < d; ++j) vec[j] /= norm;
  }
  return norm;
}

}  // namespace

std::vector<double> standardize(const double* data, size_t n, size_t d) {
  std::vector<double> out(n * d, 0.0);
  if (!data) return out;
  for (size_t j = 0; j < d; ++j) {
    double mean = 0.0;
    for (size_t i = 0; i < n; ++i) mean += data[i * d + j];
    mean /= static_cast<double>(std::max<size_t>(1, n));
    double var_sum = 0.0;
    for (size_t i = 0; i < n; ++i) {
      const double dv = data[i * d + j] - mean;
      var_sum += dv * dv;
    }
    double sd = std::sqrt(var_sum / static_cast<double>(std::max<size_t>(1, n - 1)));
    if (sd == 0.0) sd = 1.0;
    for (size_t i = 0; i < n; ++i) out[i * d + j] = (data[i * d + j] - mean) / sd;
  }
  return out;
}

PcaResult pca(const double* data, size_t n, size_t d, size_t k) {
  PcaResult out;
  if (!data || d == 0) return out;
  k = std::max<size_t>(1, std::min(k, d));
  out.n = n;
  out.d = d;
  out.k = k;

  out.mean.assign(d, 0.0);
  for (size_t i = 0; i < n; ++i) {
    for (size_t j = 0; j < d; ++j) out.mean[j] += data[i * d + j];
  }
  const double dn = static_cast<double>(std::max<size_t>(1, n));
  for (size_t j = 0; j < d; ++j) out.mean[j] /= dn;

  // Centred copy, then covariance = Xc' * Xc / (n - 1), symmetric d*d.
  std::vector<double> centered(n * d, 0.0);
  for (size_t i = 0; i < n; ++i) {
    for (size_t j = 0; j < d; ++j) centered[i * d + j] = data[i * d + j] - out.mean[j];
  }
  std::vector<double> cov(d * d, 0.0);
  const double denom = static_cast<double>(std::max<size_t>(1, n - 1));
  for (size_t i = 0; i < n; ++i) {
    const size_t row = i * d;
    for (size_t a = 0; a < d; ++a) {
      const double xa = centered[row + a];
      if (xa == 0.0) continue;
      for (size_t b = a; b < d; ++b) cov[a * d + b] += xa * centered[row + b];
    }
  }
  double total_var = 0.0;
  for (size_t a = 0; a < d; ++a) {
    for (size_t b = a; b < d; ++b) {
      const double v = cov[a * d + b] / denom;
      cov[a * d + b] = v;
      cov[b * d + a] = v;
    }
    total_var += cov[a * d + a];
  }

  out.components.assign(k * d, 0.0);
  out.explained.assign(k, 0.0);
  std::vector<double> v(d, 0.0);
  std::vector<double> w(d, 0.0);
  for (size_t c = 0; c < k; ++c) {
    // A deterministic seed that is non-degenerate and roughly orthogonal across
    // components — the reason two runs give the same embedding.
    for (size_t j = 0; j < d; ++j) {
      v[j] = std::sin(1.0 + static_cast<double>(j) * static_cast<double>(c + 1)) + 0.5;
    }
    normalize(v, d);
    double eig = 0.0;
    for (int iter = 0; iter < 256; ++iter) {
      for (size_t a = 0; a < d; ++a) {
        double s = 0.0;
        const size_t ra = a * d;
        for (size_t b = 0; b < d; ++b) s += cov[ra + b] * v[b];
        w[a] = s;
      }
      const double norm = normalize(w, d);
      double dot = 0.0;
      for (size_t a = 0; a < d; ++a) dot += w[a] * v[a];
      v = w;
      eig = norm;
      if (std::abs(std::abs(dot) - 1.0) < 1e-9) break;
    }
    for (size_t j = 0; j < d; ++j) out.components[c * d + j] = v[j];
    out.explained[c] = total_var > 0.0 ? eig / total_var : 0.0;
    // Deflate, so the next pass finds the next component rather than this one.
    for (size_t a = 0; a < d; ++a) {
      const double va = eig * v[a];
      for (size_t b = 0; b < d; ++b) cov[a * d + b] -= va * v[b];
    }
  }

  out.scores.assign(n * k, 0.0);
  for (size_t i = 0; i < n; ++i) {
    const size_t row = i * d;
    for (size_t c = 0; c < k; ++c) {
      double s = 0.0;
      const size_t cr = c * d;
      for (size_t j = 0; j < d; ++j) s += centered[row + j] * out.components[cr + j];
      out.scores[i * k + c] = s;
    }
  }
  return out;
}

}  // namespace photon::ml
